openpype/hosts/nuke/api/actions.py

```diff
@@ -20,33 +20,31 @@ class SelectInvalidAction(pyblish.api.Action):
 
     def process(self, context, plugin):
 
-        try:
-            import nuke
-        except ImportError:
-            raise ImportError("Current host is not Nuke")
-
-        errored_instances = get_errored_instances_from_context(context,
-                                                               plugin=plugin)
+        # Get the errored instances for the plug-in
+        errored_instances = get_errored_instances_from_context(
+            context, plugin)
 
         # Get the invalid nodes for the plug-ins
         self.log.info("Finding invalid nodes..")
-        invalid = list()
+        invalid_nodes = set()
         for instance in errored_instances:
-            invalid_nodes = plugin.get_invalid(instance)
+            invalid = plugin.get_invalid(instance)
+
+            if not invalid:
+                continue
 
-            if invalid_nodes:
-                if isinstance(invalid_nodes, (list, tuple)):
-                    invalid.append(invalid_nodes[0])
-                else:
-                    self.log.warning("Plug-in returned to be invalid, "
-                                     "but has no selectable nodes.")
+            select_node = instance.data.get("transientData", {}).get("node")
+            if not select_node:
+                raise RuntimeError(
+                    "No transientData['node'] found on instance: {}".format(
+                        instance)
+                )
```

Comment:
This is not relevant. This action supposes to select invalid nodes and that is exactly what it is doing. The docsting of the class is suggesting exact functionality.

Reply:
I'm sad to see this was marked "Resolved" and directly merged after - it leaves no time to discuss.

> The docsting of the class is suggesting exact functionality.

It does not, it suggest the implementation I described.

In your current implementation the static method `get_invalid` does **NOT** return the nodes that the action should select and you're not using it like that. As such, the plugin can currently not do:

```python
class Plugin:
    @staticmethod
     def get_invalid(instance):
           return [node1, node2, node3]
```

Furthermore Maya, Fusion, Houdini all have the **SelectInvalidAction** (and Nuke did too!) where it behaves like I describe above. Only now suddenly you've refactored the code in Nuke to behave differently where it just only always select the instance's node, not the `get_invalid` reported nodes. I'd argue that being consistent across multiple hosts in implementations is a big plus too - that's why I recommended that IF you wanted to change the behavior to make sure it selects the instance node to create a dedicated action to do so; which can be **greatly simplified** in code, because the Action doesn't _need_ to check the `get_invalid` call at all since the action only appears on plugins that failed anyway.

The code should have been this:
```python
import pyblish.api

from openpype.pipeline.publish import get_errored_instances_from_context
from .lib import (
    reset_selection,
    select_nodes
)


class SelectInvalidAction(pyblish.api.Action):
    """Select invalid nodes in Nuke when plug-in failed.

    To retrieve the invalid nodes this assumes a static `get_invalid()`
    method is available on the plugin.

    """
    label = "Select invalid nodes"
    on = "failed"  # This action is only available on a failed plug-in
    icon = "search"  # Icon from Awesome Icon

    def process(self, context, plugin):

        try:
            import nuke
        except ImportError:
            raise ImportError("Current host is not Nuke")

        errored_instances = get_errored_instances_from_context(context,
                                                               plugin=plugin)

        # Get the invalid nodes for the plug-ins
        self.log.info("Finding invalid nodes..")
        invalid = list()
        for instance in errored_instances:
            invalid_nodes = plugin.get_invalid(instance)

            if invalid_nodes:
                if isinstance(invalid_nodes, (list, tuple)):
                    invalid.extend(invalid_nodes)
                else:
                    self.log.warning("Plug-in returned to be invalid, "
                                     "but has no selectable nodes.")

        # Ensure unique (process each node only once)
        invalid = list(set(invalid))

        if invalid:
            self.log.info("Selecting invalid nodes: {}".format(invalid))
            reset_selection()
            select_nodes(invalid)
        else:
            self.log.info("No invalid nodes found.")



class SelectInstanceNodeAction(pyblish.api.Action):
    """Select instance node for failed plugin."""
    label = "Select instance node"
    on = "failed"  # This action is only available on a failed plug-in
    icon = "mdi.cursor-default-click"

    def process(self, context, plugin):

        # Get the errored instances for the plug-in
        errored_instances = get_errored_instances_from_context(
            context, plugin)

        # Get the invalid nodes for the plug-ins
        self.log.info("Finding invalid nodes..")
        nodes = set()
        for instance in errored_instances:
            instance_node = instance.data.get("transientData", {}).get("node")
            if not instance_node:
                raise RuntimeError(
                    "No transientData['node'] found on instance: {}".format(
                        instance
                    )
                )
            nodes.add(select_node)
            
        if nodes:
            self.log.info("Selecting instance nodes: {}".format(invalid_nodes))
            reset_selection()
            select_nodes(list(invalid_nodes))
        else:
            self.log.info("No instance nodes found.")
```